File: sites/google_kickstart/2022/5_Milk_Tea.py
# 1. find optimal binary values
# 2. make an order-index(by 1 0 difference desc)
# 3. increase a number 1 to K and apply order-index here
# 4. xor to optimal binary values
# 5. check if exists in impossible binaries

from audioop import reverse


T = int(input())
for tc in range(1, T+1):
    # num of lines, num of impossible, len of binary
    N, M, P = map(int, input().split())
    mat = []
    for i in range(N):
        mat.append(list(map(int, list(input()))))
    
    base = []
    base_bin = []
    diff = []
    diff_bin = []
    for i in range(P):
        cnt = 0
        for j in range(N):
            cnt += mat[j][i]
        base.append(cnt)
        if cnt >= (N+1)//2:
            base_bin.append(1)
        else:
            base_bin.append(0)
        diff_bin.append([cnt, N-cnt])
        diff.append(N - cnt)
    diff_idx = [(i, a) for i, a in enumerate(diff)]
    diff_idx.sort(key=lambda x:x[1], reverse=True)
    diff_idx = [i[0] for i in diff_idx]
    
    impossibles = []
    for i in range(M):
        impossibles.append(list(map(int, list(input()))))
    
    result = 0
    for i in range(0, M+1):
        tmp = list(map(int, list(bin(i)[2:].zfill(P))))
        tmp2 = [tmp[i] for i in diff_idx]
        new_bin = [0]*P
        result = 0
        for j in range(P):
            new_bin[j] = tmp2[j]^base_bin[j]
            result += diff_bin[j][new_bin[j]]
        # print(result)
        if new_bin in impossibles:
            continue
        else:
            break
        
    
    print(f"Case #{tc}: {result}")

File: sites/google_kickstart/2022/5_Milk_Tea.cpp
// 1. find optimal binary values
// 2. make an order-index(by 1 0 difference desc)
// 3. increase a number 1 to K and apply order-index here
// 4. xor to optimal binary values
// 5. check if exists in impossible binaries

#include<bits/stdc++.h>

using namespace std;

int main()
{
    int T;
    cin >> T;
    for(int tc=1; tc<T+1; ++tc)
    {
        int N, M, P;
        cin >> N>> M >> P;
        
        // find optimal values
        bool mat[100][100] = {0,};
        for(int j=0; j<M; ++j)
        {
            string s;
            cin >> s;
            for(int k=0; k<N; ++k)
            {
                mat[j][k] = s[k] - '0';
            }
        }
        
        int optimal[100] = {0,};
        for(int j=0; j<N; ++j)
        {
            int sum=0;
            for(int k=0; k<M; ++k)
            {
                sum+=mat[k][j];
            }
            optimal[j] = sum;
        }
        
        int idx[100] = {0,};
        int diff[100] = {0,};
        for(int i=0; i<N; ++i)
        {
            diff[i] = N - optimal[i];
        }
        for(int i=0; i<N; ++i)
        {
            idx[i] = i;
        }
        
        // sort by diff desc
        for(int i=0; i<N-1; ++i)
        {
            for(int j=i+1; j<N; ++j)
            {
                if(diff[i] < diff[j])
                {
                    int tmp;
                    tmp = diff[j];
                    diff[j] = diff[i];
                    diff[i] = tmp;
                    
                    tmp = idx[j];
                    idx[j] = idx[i];
                    idx[i] = tmp;
                }
            }
        }
        
        bool base[100] = {0,};
        for(int i=0; i<N; ++i)
        {
            if(optimal[i] >= (N+1)/2)
            {
                base[i] = 1;
            }else
            {
                base[i] = 0;
            }
        }
        
        vector<string> vs;
        for(int i=0; i<P; ++i)
        {
            string tmp;
            cin >> tmp;
            vs.push_back(tmp);
        }
        
        for(int i=0; i<P; ++i)
        {
            
        }
        
        
        
        
        
        cout << "Case #" << tc << ": " << result << "\n";
    }
    return 0;
}